File: threedollar-api-user-service/src/main/java/com/depromeet/threedollar/api/userservice/service/review/dto/response/ReviewInfoResponse.java
package com.depromeet.threedollar.api.userservice.service.review.dto.response;

import com.depromeet.threedollar.api.core.service.common.dto.response.AuditingTimeResponse;
import com.depromeet.threedollar.domain.rds.domain.userservice.review.Review;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@ToString
@Getter
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ReviewInfoResponse extends AuditingTimeResponse {

    private Long reviewId;

    private Long storeId;

    @Nullable
    private String contents;

    private int rating;

    @Builder(access = AccessLevel.PRIVATE)
    private ReviewInfoResponse(Long reviewId, Long storeId, @Nullable String contents, int rating) {
        this.reviewId = reviewId;
        this.storeId = storeId;
        this.contents = contents;
        this.rating = rating;
    }

    public static ReviewInfoResponse of(@NotNull Review review) {
        ReviewInfoResponse response = ReviewInfoResponse.builder()
            .reviewId(review.getId())
            .storeId(review.getStoreId())
            .contents(review.getContents())
            .rating(review.getRating())
            .build();
        response.setAuditingTimeByEntity(review);
        return response;
    }

}

File: threedollar-api-admin-service/src/main/kotlin/com/depromeet/threedollar/api/adminservice/service/userservice/review/dto/response/ReviewInfoResponse.kt
package com.depromeet.threedollar.api.adminservice.service.userservice.review.dto.response

import com.depromeet.threedollar.api.adminservice.service.userservice.store.dto.response.StoreInfoResponse
import com.depromeet.threedollar.api.core.common.dto.AuditingTimeResponse
import com.depromeet.threedollar.domain.rds.domain.userservice.review.Review
import com.depromeet.threedollar.domain.rds.domain.userservice.store.Store
import com.depromeet.threedollar.domain.rds.domain.userservice.user.User

data class ReviewInfoResponse(
    val reviewId: Long,
    val contents: String,
    val rating: Int,
    val user: UserInfoResponse?,
    val store: StoreInfoResponse?,
) : AuditingTimeResponse() {

    companion object {
        fun of(review: Review, store: Store?, user: User?): ReviewInfoResponse {
            val response = ReviewInfoResponse(
                reviewId = review.id,
                contents = review.contents,
                rating = review.rating,
                user = user?.let { UserInfoResponse.of(it) },
                store = store?.let { StoreInfoResponse.of(it) }
            )
            response.setAuditingTimeByEntity(review)
            return response
        }
    }

}


